### ISSUE TITLE
imsaisim webfrontend LIB: not working on xfs filesystem

### ISSUE THREAD

**Issue by signals42** [udo-munk/z80pack#76]
After installing imsaisim, I was unable to get the disk library functionality in the web frontend working. The LIB: window in the web interface is always blank. I can upload disk images and they appear on the file system, but they never show up in the window.

I'm running CentOS 7 and have xfs file systems. It seems that xfs behaves differently from ext4 when it comes to dirent. The d_type returned for regular files is DT_UNKNOWN, not DT_REG, so on line 276 of netsrv.c all of the files in my library are filtered out:

`if (pDirent->d_type==DT_REG) {`

After switching to this, it works properly on my system:

`if (pDirent->d_type==DT_REG || pDirent->d_type==DT_UNKNOWN) {`

**Comment by udo-munk** [comment]
Linux manual page for readdir(3) says:

Currently, only some filesystems (among them: Btrfs, ext2,
ext3, and ext4) have full support for returning the file type
in d_type.  All applications must properly handle a return of
DT_UNKNOWN.

Question is what does the function return for a directory, or pipe, or ... Is it always DT_UNKNOWN?

**Comment by signals42** [comment]
According to this page,  http://northernmost.org/blog/readdir-and-directories-on-xfs/index.html it's always unknown and the fix is probably to use stat(). I may not have a chance for a day or two, but I'll test it on my xfs system and let you know if i am also seeing DT_UNKNOWN for everything.Sent from my Samsung Galaxy smartphone.
-------- Original message --------From: udo-munk <notifications@github.com> Date: 11/18/19  11:06 AM  (GMT-05:00) To: udo-munk/z80pack <z80pack@noreply.github.com> Cc: signals42 <signals42@gmail.com>, Author <author@noreply.github.com> Subject: Re: [udo-munk/z80pack] imsaisim webfrontend LIB: not working on xfs filesystem (#76) Linux manual page for readdir(3) says:
Currently, only some filesystems (among them: Btrfs, ext2,
ext3, and ext4) have full support for returning the file type
in d_type.  All applications must properly handle a return of
DT_UNKNOWN.
Question is what does the function return for a directory, or pipe, or ... Is it always DT_UNKNOWN?

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/udo-munk/z80pack/issues/76?email_source=notifications\u0026email_token=AANILXRHHM2FKGGT435SFZDQUK4RXA5CNFSM4JOUSB5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEK6TEI#issuecomment-555084177",
"url": "https://github.com/udo-munk/z80pack/issues/76?email_source=notifications\u0026email_token=AANILXRHHM2FKGGT435SFZDQUK4RXA5CNFSM4JOUSB5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEK6TEI#issuecomment-555084177",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by udo-munk** [comment]
OK, that's what I guessed too, so this fix might screw up other file types. I have to setup a Linux VM with xfs to have a closer look, I also need some time for this, currently all my VM's use ext4.

**Comment by signals42** [comment]
You can also do something like:

```
dd bs=1024 count=102400 if=/dev/zero of=./xfs.filesystem
mkfs.xfs ./xfs.filesystem
sudo mount -o loop ./xfs.filesystem /mnt/xfs
```

If you don't want to do a full VM for testing.

Also, I have confirmed that all files, directories, and fifos come back as DT_UNKNOWN on xfs. Strangely "." and ".." are DT_DIR, but everything else is DT_UNKNOWN.

**Comment by udo-munk** [comment]
Yep, or I could add another virtual disk and put a xfs partition on that. Anyway, got latest CentOS 8 and did a full VM install to have a look at this. The install created a xfs for / filesystem, so I will be able to reproduce this problem and help finding a workarround for your problem.

**Comment by EtchedPixels** [comment]
> Question is what does the function return for a directory, or pipe, or ... Is it always DT_UNKNOWN?

It could return DT_UNKNOWN or the expected DT_ type. If it's DT_UNKNOWN stat it if you need to know. It's just that some file systems don't have the meta data laid out so that turning a name into a type is always free and not everyone wants to know.

**Comment by signals42** [comment]
Let me start out by saying that I'm not a native C speaker, and any training I had in it was decades ago in college, so expect this to be terrible. Anyway, I added the following 

```
    struct stat sb;
    char fullpath[PATH_MAX];
```

to DirectoryHandler(). Then I changed the directory scan logic as so:

```
            while ((pDirent = readdir(pDir)) != NULL) {
                LOGD(TAG, "GET directory: %s type: %d", pDirent->d_name, pDirent->d_type);
                snprintf(&fullpath, PATH_MAX, "%s/%s", (char *)path, pDirent->d_name);
                if (stat(fullpath,&sb) != -1 && S_ISREG(sb.st_mode)) {
                        httpdPrintf(conn, "%c\"%s\"", (i++ > 0)?',':' ', pDirent->d_name);
                }
            }
```

It seems to work, and it is my understanding that stat() should be reliable no matter what file system we are using.

**Comment by udo-munk** [comment]
It isn't terrible, pretty much OK and unlike your first approach it won't cause problems. I've tested this fix with OSX and several Linux version on ext4 and xfs, works ok. Correct, stat() will return the file type, no matter which filesystem.

Your fix is comitted with 28d742e, thanks for finding this, so far I was not aware about the problem.

**Comment by udo-munk** [comment]
Also works on Windows with ntfs and vfat.
